Test failure in CI build 1232

The following test appears to have failed:

[#1232](https://circleci.com/gh/cockroachdb/cockroach/1232):

```
I0408 19:29:24.693501     270 multiraft.go:638] HardState updated: {Term:6 Vote:257 Commit:35 XXX_unrecognized:[]}
I0408 19:29:24.694914     270 multiraft.go:641] New Entry[0]: 6/35 EntryNormal 13d3209ccf94d81865157426e50ec28b: raft_id:1 cmd:<end_transaction:<header:<timestamp:<wall_time:0 logical:31 > cmd_id:<wall_time:1428521364676204568 random:7283855682742174347 > key:"b
I0408 19:29:24.696199     270 multiraft.go:644] Committed Entry[0]: 6/35 EntryNormal 13d3209ccf94d81865157426e50ec28b: raft_id:1 cmd:<end_transaction:<header:<timestamp:<wall_time:0 logical:31 > cmd_id:<wall_time:1428521364676204568 random:7283855682742174347 > key:"b
I0408 19:29:24.704034     270 raft.go:390] raft: 101 became follower at term 5
I0408 19:29:24.704266     270 raft.go:207] raft: newRaft 101 [peers: [101], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x10 pc=0xb00a56]

goroutine 1674 [running]:
github.com/cockroachdb/cockroach/multiraft.(*state).propose(0xc2080a0b70, 0xc208656900)
	/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:625 +0xc6
github.com/cockroachdb/cockroach/multiraft.(*state).start(0xc2080a0b70)
	/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:477 +0x1ad7
created by github.com/cockroachdb/cockroach/multiraft.(*MultiRaft).Start
	/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:149 +0x60

--
goroutine 1675 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).processRaft(0xc208398160)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1121 +0xee6
created by github.com/cockroachdb/cockroach/storage.(*Store).Start
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:477 +0xc07
FAIL	github.com/cockroachdb/cockroach/storage	3.446s
=== RUN TestBatchBasics
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
--- PASS: TestBatchMerge (0.00s)
=== RUN TestBatchProto
--- PASS: TestBatchProto (0.00s)
=== RUN TestBatchScan
--- PASS: TestBatchScan (0.00s)
I0408 19:29:24.693501     270 multiraft.go:638] HardState updated: {Term:6 Vote:257 Commit:35 XXX_unrecognized:[]}
I0408 19:29:24.694914     270 multiraft.go:641] New Entry[0]: 6/35 EntryNormal 13d3209ccf94d81865157426e50ec28b: raft_id:1 cmd:<end_transaction:<header:<timestamp:<wall_time:0 logical:31 > cmd_id:<wall_time:1428521364676204568 random:7283855682742174347 > key:"b
I0408 19:29:24.696199     270 multiraft.go:644] Committed Entry[0]: 6/35 EntryNormal 13d3209ccf94d81865157426e50ec28b: raft_id:1 cmd:<end_transaction:<header:<timestamp:<wall_time:0 logical:31 > cmd_id:<wall_time:1428521364676204568 random:7283855682742174347 > key:"b
I0408 19:29:24.704034     270 raft.go:390] raft: 101 became follower at term 5
I0408 19:29:24.704266     270 raft.go:207] raft: newRaft 101 [peers: [101], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x10 pc=0xb00a56]

goroutine 1674 [running]:
github.com/cockroachdb/cockroach/multiraft.(*state).propose(0xc2080a0b70, 0xc208656900)
	/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:625 +0xc6
github.com/cockroachdb/cockroach/multiraft.(*state).start(0xc2080a0b70)
	/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:477 +0x1ad7
created by github.com/cockroachdb/cockroach/multiraft.(*MultiRaft).Start
	/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:149 +0x60

--
goroutine 1675 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).processRaft(0xc208398160)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1121 +0xee6
created by github.com/cockroachdb/cockroach/storage.(*Store).Start
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:477 +0xc07
FAIL	github.com/cockroachdb/cockroach/storage	3.446s
=== RUN TestBatchBasics
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
--- PASS: TestBatchMerge (0.00s)
=== RUN TestBatchProto
--- PASS: TestBatchProto (0.00s)
=== RUN TestBatchScan
--- PASS: TestBatchScan (0.00s)

```
Please assign, take a look and update the issue accordingly.

--- a/kv/dist_sender.go
+++ b/kv/dist_sender.go
@@ -111,6 +111,7 @@ type DistSender struct {
 	// outside of tests.
 	rpcSend         rpcSendFn
 	rpcRetryOptions util.RetryOptions
+	closer          chan struct{}
 }
 
 // rpcSendFn is the function type used to dispatch RPC calls.
@@ -152,6 +153,7 @@ func NewDistSender(ctx *DistSenderContext, gossip *gossip.Gossip) *DistSender {
 	ds := &DistSender{
 		clock:  clock,
 		gossip: gossip,
+		closer: make(chan struct{}),
 	}
 	ds.nodeID = ctx.NodeID
 	rcSize := ctx.RangeDescriptorCacheSize
@@ -429,6 +431,14 @@ func (ds *DistSender) Send(call *client.Call) {
 	for {
 		reply := call.Reply
 		err := util.RetryWithBackoff(retryOpts, func() (util.RetryStatus, error) {
+			select {
+			case <-ds.closer:
+				call.Reply.Header().SetGoError(
+					proto.NewTransactionRetryError(call.Args.Header().Txn))
+				return util.RetryBreak, nil
+			default:
+			}
+
 			reply.Header().Reset()
 			descNext = nil
 			desc, err := ds.rangeCache.LookupRangeDescriptor(args.Header().Key)
@@ -544,6 +554,7 @@ func (ds *DistSender) Send(call *client.Call) {
 	}
 }
 
-// Close implements the client.KVSender interface. It's a noop for the
-// distributed sender.
-func (ds *DistSender) Close() {}
+// Close implements the client.KVSender interface.
+func (ds *DistSender) Close() {
+	close(ds.closer)
+}

--- a/kv/local_sender.go
+++ b/kv/local_sender.go
@@ -31,12 +31,16 @@ import (
 type LocalSender struct {
 	mu       sync.RWMutex                     // Protects storeMap and addrs
 	storeMap map[proto.StoreID]*storage.Store // Map from StoreID to Store
+	closer   chan struct{}
 }
 
 // NewLocalSender returns a local-only sender which directly accesses
 // a collection of stores.
 func NewLocalSender() *LocalSender {
-	return &LocalSender{storeMap: map[proto.StoreID]*storage.Store{}}
+	return &LocalSender{
+		storeMap: map[proto.StoreID]*storage.Store{},
+		closer:   make(chan struct{}),
+	}
 }
 
 // GetStoreCount returns the number of stores this node is exporting.
@@ -104,6 +108,13 @@ func (ls *LocalSender) Send(call *client.Call) {
 		MaxAttempts: 2,
 	}
 	util.RetryWithBackoff(retryOpts, func() (util.RetryStatus, error) {
+		select {
+		case <-ls.closer:
+			call.Reply.Header().SetGoError(
+				proto.NewTransactionRetryError(call.Args.Header().Txn))
+			return util.RetryBreak, nil
+		default:
+		}
 		call.Reply.Header().Error = nil
 		var err error
 		var store *storage.Store
@@ -168,6 +179,7 @@ func (ls *LocalSender) Send(call *client.Call) {
 func (ls *LocalSender) Close() {
 	ls.mu.RLock()
 	defer ls.mu.RUnlock()
+	close(ls.closer)
 	for _, store := range ls.storeMap {
 		store.Stop()
 	}

--- a/multiraft/multiraft.go
+++ b/multiraft/multiraft.go
@@ -537,6 +537,21 @@ func (s *state) coalescedHeartbeat() {
 func (s *state) stop() {
 	log.V(6).Infof("node %v stopping", s.nodeID)
 	s.writeTask.stop()
+
+	// Drain the create/remove group channels because other threads may be blocking
+	// on these operations.
+	done := false
+	for !done {
+		select {
+		case op := <-s.createGroupChan:
+			op.ch <- util.Errorf("shutting down")
+		case op := <-s.removeGroupChan:
+			op.ch <- util.Errorf("shutting down")
+		default:
+			done = true
+		}
+	}
+
 	s.stopper.SetStopped()
 }
 

--- a/storage/range.go
+++ b/storage/range.go
@@ -578,11 +578,20 @@ func (r *Range) addReadWriteCmd(method string, args proto.Request, reply proto.R
 	// Create a completion func for mandatory cleanups which we either
 	// run synchronously if we're waiting or in a goroutine otherwise.
 	completionFunc := func() error {
+		var err error
 		// First wait for raft to commit or abort the command.
-		err := <-raftChan
+		select {
+		case err = <-raftChan:
+		case <-r.closer:
+			err = util.Errorf("range is closing")
+		}
 		if err == nil {
 			// Next if the command was commited, wait for the range to apply it.
-			err = <-pendingCmd.done
+			select {
+			case err = <-pendingCmd.done:
+			case <-r.closer:
+				err = util.Errorf("range is closing")
+			}
 		}
 
 		// As for reads, update timestamp cache with the timestamp

--- a/storage/store.go
+++ b/storage/store.go
@@ -1079,7 +1079,9 @@ func (s *Store) ProposeRaftCommand(idKey cmdIDKey, cmd proto.InternalRaftCommand
 	// Lazily create group. TODO(bdarnell): make this non-lazy
 	err := s.multiraft.CreateGroup(uint64(cmd.RaftID))
 	if err != nil {
-		log.Fatal(err)
+		ch := make(chan error, 1)
+		ch <- err
+		return ch
 	}
 
 	data, err := gogoproto.Marshal(&cmd)

--- a/util/leaktest/leaktest.go
+++ b/util/leaktest/leaktest.go
@@ -57,11 +57,7 @@ func interestingGoroutines() (gs []string) {
 			strings.Contains(stack, "created by runtime.gc") ||
 			strings.Contains(stack, "github.com/cockroachdb/cockroach/util/leaktest.interestingGoroutines") ||
 			strings.Contains(stack, "runtime.MHeap_Scavenger") ||
-			strings.Contains(stack, "golang/glog.init") ||
-			// Fire-and-forget write intent cleanups. These end up deadlocked
-			// because the store is already stopped by the time they run.
-			// TODO(bdarnell): clean these up better.
-			strings.Contains(stack, "*txnMetadata).close") {
+			strings.Contains(stack, "golang/glog.init") {
 			continue
 		}
 		gs = append(gs, stack)

